<a href="https://colab.research.google.com/github/SharonEstivalez/FisicaComputacional1/blob/master/Actividad_3_Sharon.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ***$$ACTIVIDAD  \space 3$$***



Pacheco Rodriguez Sharon Estivalez

 27/ 01/ 21
 


In [30]:
#Importar librerías
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [31]:
#Importar el archivo 
#df = pd.read_csv("https://raw.githubusercontent.com/SharonEstivalez/FisicaComputacional1/master/Actividad%203/10090.csv")
#df = df.drop(range(0,16))

#Tomar url en raw
url = 'https://raw.githubusercontent.com/SharonEstivalez/FisicaComputacional1/master/Actividad%203/10090.txt'

#Nombres para las columnas
col =  ['Fecha', 'Precip', 'Evap', 'Tmax', 'Tmin'] 

# encoding='cp1251' Windows (Europa Occidental)
# sep='\s+' el separador de datos es de uno o más espacios
#header=None no hay encabezado
#skiprows=19 Se brincará los primeros 19 renglones 
#skipfooter=1 Se brincará el último renglón
#engine='python' Se procesan los datos utilizando Python 
df_act = pd.read_csv(url, names=col, encoding='cp1251', sep='\s+', header=None, skiprows=19, skipfooter=1, engine='python')

# Elimina los renglones que tengan algún número nulo.  
#df = df.dropna()



In [32]:
df_act.head()
#df_act.tail()


Fecha Precip Evap  Tmax Tmin
0  01/08/1963      0  3.6    28   10
1  02/08/1963      0  4.8    21   11
2  03/08/1963      0  4.5    25   10
3  04/08/1963      0    4    18   11
4  05/08/1963      9  2.5  24.5   12

In [33]:
# Revisamos el encabezado y el final del dataframe
print(df_act.head(10))  # Mostrar y enlistar los primeros 5 renglones por default. ## Si se desea imprimir n renglones: df.head(n)
print(df_act.tail(10))  #Igual pero con los últimos


        Fecha Precip Evap  Tmax Tmin
0  01/08/1963      0  3.6    28   10
1  02/08/1963      0  4.8    21   11
2  03/08/1963      0  4.5    25   10
3  04/08/1963      0    4    18   11
4  05/08/1963      9  2.5  24.5   12
5  06/08/1963      0  3.3    26   12
6  07/08/1963      0  2.8    20   12
7  08/08/1963      0  2.8    20   12
8  09/08/1963      7  3.5    26   13
9  10/08/1963   13.5  4.4    26   12
            Fecha Precip  Evap Tmax Tmin
19582  22/12/2018      0  Nulo   21   -4
19583  23/12/2018      0  Nulo   20   -3
19584  24/12/2018      0  Nulo   21   -4
19585  25/12/2018      0  Nulo   21    2
19586  26/12/2018      0  Nulo   23    2
19587  27/12/2018      0  Nulo   22    2
19588  28/12/2018      0  Nulo   22   -3
19589  29/12/2018      0  Nulo   20    1
19590  30/12/2018      0  Nulo   20   -1
19591  31/12/2018      0  Nulo   22   -3


In [34]:
# Investigamos su tamaño con la función shape (renglones, columnas)
df_act.shape

(19592, 5)

In [35]:
#Información del DataFrame: número de renglones, columnas, nombres, tipo de datos, número de entradas no nulas, y otros
df_act.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19592 entries, 0 to 19591
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Fecha   19592 non-null  object
 1   Precip  19592 non-null  object
 2   Evap    19592 non-null  object
 3   Tmax    19592 non-null  object
 4   Tmin    19592 non-null  object
dtypes: object(5)
memory usage: 765.4+ KB


In [36]:
#Hacer una copia del DataFrame
df_sp0 = df_act.copy()

#df_sp0.head() #Mis valores nulos están en los datos del final
df_sp0.tail()


Fecha Precip  Evap Tmax Tmin
19587  27/12/2018      0  Nulo   22    2
19588  28/12/2018      0  Nulo   22   -3
19589  29/12/2018      0  Nulo   20    1
19590  30/12/2018      0  Nulo   20   -1
19591  31/12/2018      0  Nulo   22   -3

In [37]:
#Reemplazar "Nulo" por espacio en blanco
str_Nulo = 'Nulo'
df_sp1 = df_sp0.replace(to_replace=str_Nulo, value='', regex=True)
df_sp1.tail()


Fecha Precip Evap Tmax Tmin
19587  27/12/2018      0        22    2
19588  28/12/2018      0        22   -3
19589  29/12/2018      0        20    1
19590  30/12/2018      0        20   -1
19591  31/12/2018      0        22   -3

In [38]:
# Mostrar información actualizada del DataFrame
df_sp1.info()
print(df_sp1.tail())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19592 entries, 0 to 19591
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Fecha   19592 non-null  object
 1   Precip  19592 non-null  object
 2   Evap    19592 non-null  object
 3   Tmax    19592 non-null  object
 4   Tmin    19592 non-null  object
dtypes: object(5)
memory usage: 765.4+ KB
            Fecha Precip Evap Tmax Tmin
19587  27/12/2018      0        22    2
19588  28/12/2018      0        22   -3
19589  29/12/2018      0        20    1
19590  30/12/2018      0        20   -1
19591  31/12/2018      0        22   -3


# *convertir a número flotante o numérico los datos de Precipitación, Evaporación, Temperatura Máxima y Temperatura mínima utilizando la función: df.to_numeric()*

In [39]:

#Convertir las valores de Precip, Evap, Tmax y Tmin en flotantes
cols_list = ['Precip', 'Evap', 'Tmax', 'Tmin']
for cols in cols_list:
  df_sp1[cols] = pd.to_numeric(df_sp1[cols], errors='coerce')

df_sp1.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19592 entries, 0 to 19591
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Fecha   19592 non-null  object 
 1   Precip  19558 non-null  float64
 2   Evap    17688 non-null  float64
 3   Tmax    19484 non-null  float64
 4   Tmin    19487 non-null  float64
dtypes: float64(4), object(1)
memory usage: 765.4+ KB


In [ ]:
#Hay valores faltantes, se nota en los que tiene la fecha; así que se suman los nulos. 
df_sp1.isnull().sum()   # Regresa la suma de número faltantes de cada variable. 


Fecha        0
Precip      34
Evap      1904
Tmax       108
Tmin       105
dtype: int64

In [40]:
#Mostrar encabezado y final del DataFrame
print(df_sp1.head())
print(df_sp1.tail())

        Fecha  Precip  Evap  Tmax  Tmin
0  01/08/1963     0.0   3.6  28.0  10.0
1  02/08/1963     0.0   4.8  21.0  11.0
2  03/08/1963     0.0   4.5  25.0  10.0
3  04/08/1963     0.0   4.0  18.0  11.0
4  05/08/1963     9.0   2.5  24.5  12.0
            Fecha  Precip  Evap  Tmax  Tmin
19587  27/12/2018     0.0   NaN  22.0   2.0
19588  28/12/2018     0.0   NaN  22.0  -3.0
19589  29/12/2018     0.0   NaN  20.0   1.0
19590  30/12/2018     0.0   NaN  20.0  -1.0
19591  31/12/2018     0.0   NaN  22.0  -3.0


In [42]:
# Estadística básica de las variables numéricas a 2 cifras significativas
df_sp1.describe().round(2)


Precip      Evap      Tmax      Tmin
count  19558.00  17688.00  19484.00  19487.00
mean       1.55      5.78     25.11      6.95
std        5.31      2.66      4.92      5.50
min        0.00      0.10     -8.00    -13.00
25%        0.00      4.00     22.00      3.00
50%        0.00      5.50     25.00      8.00
75%        0.00      7.40     28.50     12.00
max       70.00     18.00     45.00     20.00

**Interpretación de los resultados anteriores**

Se puede notar que las precipitaciones varían drásticamente en el mínimo y máximo, de 0mm a 7mm, con un promedio de 1.55mm. 
La evaporación tiende a 5.78mm en promedio, variando sus mínimos y máximos de 
0.10mm a 18mm.
En las temperaturas máximas registradas se encuentra un promedio de 25.11°C, con mínima de -8°C y máxima de 45°C.
En las temperaturas mínimas registradas se encuentra un promedio de 6.95°C, una mínima de -13°C y una máxima de 20°C.
Con esto se puede notar que las temperaturas cambian drásticamente tanto en mínimas como máximas, al igual que la variación de las precipitaciones. 

# *Fecha*

In [43]:
#Hacer copia del DataFrame anterior para trabajar con ella

df_sp2 = df_sp1.copy()

df_sp2['Fecha'].head()

0    01/08/1963
1    02/08/1963
2    03/08/1963
3    04/08/1963
4    05/08/1963
Name: Fecha, dtype: object

In [49]:


#Tomar "Fecha" y convertirlo al formato fecha que entiende python
df_sp2['Fecha'] = pd.to_datetime(df_sp2['Fecha'], dayfirst=True).copy()
print(df_sp2.head())


       Fecha  Precip  Evap  Tmax  Tmin
0 1963-08-01     0.0   3.6  28.0  10.0
1 1963-08-02     0.0   4.8  21.0  11.0
2 1963-08-03     0.0   4.5  25.0  10.0
3 1963-08-04     0.0   4.0  18.0  11.0
4 1963-08-05     9.0   2.5  24.5  12.0


In [51]:
#Confirmar que las variables sean del tipo que queremos

df_sp2.dtypes

Fecha     datetime64[ns]
Precip           float64
Evap             float64
Tmax             float64
Tmin             float64
dtype: object

In [52]:
#Añadir dos columnas, año y fecha

df_sp2['Año'] = df_sp2['Fecha'].dt.year
df_sp2['Mes'] = df_sp2['Fecha'].dt.month


In [55]:
#Verificar que se añadió el cambio anterior, que ya están las columnas nuevas
print(df_sp2.head())
print(df_sp2.tail())


       Fecha  Precip  Evap  Tmax  Tmin   Año  Mes
0 1963-08-01     0.0   3.6  28.0  10.0  1963    8
1 1963-08-02     0.0   4.8  21.0  11.0  1963    8
2 1963-08-03     0.0   4.5  25.0  10.0  1963    8
3 1963-08-04     0.0   4.0  18.0  11.0  1963    8
4 1963-08-05     9.0   2.5  24.5  12.0  1963    8
           Fecha  Precip  Evap  Tmax  Tmin   Año  Mes
19587 2018-12-27     0.0   NaN  22.0   2.0  2018   12
19588 2018-12-28     0.0   NaN  22.0  -3.0  2018   12
19589 2018-12-29     0.0   NaN  20.0   1.0  2018   12
19590 2018-12-30     0.0   NaN  20.0  -1.0  2018   12
19591 2018-12-31     0.0   NaN  22.0  -3.0  2018   12


In [57]:
# Ver que las nuevas columnas, año y mes, son variables enteras 
df_sp2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19592 entries, 0 to 19591
Data columns (total 7 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   Fecha   19592 non-null  datetime64[ns]
 1   Precip  19558 non-null  float64       
 2   Evap    17688 non-null  float64       
 3   Tmax    19484 non-null  float64       
 4   Tmin    19487 non-null  float64       
 5   Año     19592 non-null  int64         
 6   Mes     19592 non-null  int64         
dtypes: datetime64[ns](1), float64(4), int64(2)
memory usage: 1.0 MB
